### PR TITLE
[Torch] Use FP32 opmath for FP16 batch norm inference

### PR DESCRIPTION
**Pull request by gokcan-karakus**
Summary

Use FP32 opmath when decomposing FP16 batch normalization in inference mode.

Performing the batch-normalization arithmetic directly in FP16 can introduce excessive numerical error. This change converts the input, running statistics, weight, and bias to FP32 before performing the inference calculation, then converts the final output back to FP16.

The change applies only to FP16 inference. Training behavior and other input dtypes are unchanged.

Testing

- Added a decomposition lit test verifying that FP16 inference operands are converted to FP32 and the result is converted back to FP16.
- Added `BatchNormInferenceF16Module_basic`.
- Enabled and passed the test for the TOSA configuration.
- Passed focused batch-normalization regression tests.
- Passed the native lit test suite.
- Recorded the existing StableHLO limitation in its expected crashing set.
- Passed pre-commit checks.

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -8624,6 +8624,19 @@ class DecomposeAtenNativeBatchNormOp
         return rewriter.notifyMatchFailure(op, "expected bias to be rank 1");
     }
 
+    Type originalDtype = cast<ValueTensorType>(input.getType()).getDtype();
+    bool useF32Opmath = !training && originalDtype.isF16();
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:8731]: I think PT does fp32 compute for bf16 type too.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -8624,6 +8624,19 @@ class DecomposeAtenNativeBatchNormOp
         return rewriter.notifyMatchFailure(op, "expected bias to be rank 1");
     }
 
+    Type originalDtype = cast<ValueTensorType>(input.getType()).getDtype();
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Use `getOptionalDtype` instead of `getDtype` as a safety check similar to other usage in this code-base.

File: test/Dialect/Torch/decompose-complex-ops.mlir [hunk 3]
```diff
@@ -1256,3 +1256,25 @@ func.func @torch.aten.diag_2d(%arg0: !torch.vtensor<[3,4],f32>) -> !torch.vtenso
   %0 = torch.aten.diag %arg0, %int0 : !torch.vtensor<[3,4],f32>, !torch.int -> !torch.vtensor<[3],f32>
   return %0 : !torch.vtensor<[3],f32>
 }
+
+// -----
+
+// CHECK-LABEL: func.func @batch_norm_fp16_opmath
+// CHECK-COUNT-5: torch.aten.to.dtype
+// CHECK: torch.aten.rsqrt {{.*}} -> !torch.vtensor<[1,2,1],f32>
```
Comment (sahas3) [inline test/Dialect/Torch/decompose-complex-ops.mlir]: Can you add a // CHECK-NOT: torch.aten.to.dtype before rsqrt, so that in future if a change starts producing >5 dtype the LIT test will catch it?

**Review by sahas3 — CHANGES_REQUESTED** [review]
Thanks for the patch, left a few suggestions.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 1 above)
Comment (gokcan-karakus) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:8731]: Thank you very much.
PyTorch also uses FP32 opmath for BF16 inference. I investigated extending this PR accordingly, but the equivalent BF16 `fx_importer_tosa` e2e test currently fails at the RefBackend boundary because BF16 memref arguments are unsupported. Since added behavior requires passing e2e coverage, I have kept this PR scoped to FP16. BF16 opmath should be added separately once the BF16 runtime/ABI path supports end-to-end execution.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 2 above)
Comment (gokcan-karakus) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Addressed. The decomposition now uses `getOptionalDtype()` and reports a match failure when the input dtype is unavailable.

File: test/Dialect/Torch/decompose-complex-ops.mlir (hunk 3 above)
Comment (gokcan-karakus) [reply test/Dialect/Torch/decompose-complex-ops.mlir]: Addressed. I added `CHECK-NOT: torch.aten.to.dtype` between the five expected input promotions and `rsqrt`, so an unexpected additional cast before `rsqrt` will fail the regression.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -8724,6 +8724,22 @@ class DecomposeAtenNativeBatchNormOp
         return rewriter.notifyMatchFailure(op, "expected bias to be rank 1");
     }
 
+    Type originalDtype =
+        cast<ValueTensorType>(input.getType()).getOptionalDtype();
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: I think you need to check for `getOptionalDtype` not returning `null` before you cast here.

**Review by sahas3 — APPROVED** [review]
LG!

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 1 above)
Comment (sahas3) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:8731]: Thanks for looking into this, makes sense to keep this for fp16 for now.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 4 above)
Comment (gokcan-karakus) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Addressed. The decomposition now safely obtains a BaseTensorType and verifies that its dtype is known before accessing it. Inputs without a known dtype return a match failure instead of reaching an unchecked cast or dtype access.
